### PR TITLE
bugfix: use same memory in ringbuffer

### PR DESCRIPTION
**Pull request by fuweid**
Signed-off-by: Wei Fu <fuweid89@gmail.com>

### Ⅰ. Describe what this PR did

ContainerIO' backend will read data into ringbuffer. But the data blocks
use the same memory so that the comming data will override the previous
one.

### Ⅱ. Does this pull request fix one issue?

When the input is too long, ContainerIO will separate one line into two data blocks. But the blocks shares the same memory, the new one will override the old one, like 

```
# vagrant @ ubuntu-xenial in ~/go/src/github.com/alibaba/pouch/test on git:bugfix_conflict_data_in_ringbuffer o [14:51:15]
$ sudo pouch run -d busybox:1.28 sleep 10000
08921f5002032b14763eceb90ad4bbbfb25a352e819c9c0ba1aaab54bb68b5bf

# vagrant @ ubuntu-xenial in ~/go/src/github.com/alibaba/pouch/test on git:bugfix_conflict_data_in_ringbuffer o [14:51:21]
$ sudo pouch exec -i 08921f5002032b14763eceb90ad4bbbfb25a352e819c9c0ba1aaab54bb68b5bf sh
ksjfjdsfjklsdjkfljsdkljfklsjfjsdjfjskjfksfjdksjfjsjdfjksjfkjskdjfkjsdkfjksdjfkjskdjfksdjkfjsdkfjksdjfkjsdkfjksdjfkjsdkfjsdfjsk
sh: ksjfjdsfjklsdjkfljsdkljfklsjfjsdjfjskjfksfjdksjfjsjdfjksjfkjskdjfkjsdkfjksdjfkjskdjfksdjkfjsdkfjksdjfkjsdkfjksdjfkjsdkfjsdfjsk: not found
dkfjksjfsdkfjksjdkjfkjsdkfjkjdkjfkdsjffjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj
sh: jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj: not found
sh: jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj: not found
exit
```

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

Added

### Ⅳ. Describe how to verify it

input >128 chars
```
# vagrant @ ubuntu-xenial in ~/go/src/github.com/alibaba/pouch on git:bugfix_conflict_data_in_ringbuffer o [14:55:15]
$ sudo pouch exec -i 08921f5002032b14763eceb90ad4bbbfb25a352e819c9c0ba1aaab54bb68b5bf sh
dkfjksjfsdkfjksjdkjfkjsdkfjkjdkjfkdsjffjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj
sh: dkfjksjfsdkfjksjdkjfkjsdkfjkjdkjfkdsjffjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj: not found
exit
```

### Ⅴ. Special notes for reviews

I hate the pattern...

### CODE REVIEW

File: daemon/containerio/container_io.go [hunk 1]
```diff
@@ -356,7 +356,16 @@ func (cio *ContainerIO) converge(name, id string, in io.Reader) {
...
 		}
 
-		cover, err := cio.ring.Push(data[:n])
+		// FIXME(fuwei): In case that the data slice is reused by the writer,
+		// we should copy the data before we push it into the ringbuffer.
+		// The previous data shares the same address with the coming data.
+		// If we don't copy the data and the previous data isn't consumed by
+		// ringbuf pop action, the incoming data will override the previous data
+		// in the ringbuf.
+		copyData := make([]byte, n)
```
Comment (Ace-Tang) [inline daemon/containerio/container_io.go]: From the comment, `Copy` make data not be override, right?

Comment (fuweid) [reply daemon/containerio/container_io.go]: Yes

Comment (Ace-Tang) [reply daemon/containerio/container_io.go]: More qeustion, as I do not know all ringbuf logic. The data will always append if pop not assume?

Comment (fuweid) [reply daemon/containerio/container_io.go]: yes. if the size reaches the capacity, ringbuffer will drop the old one.

**Comment by Ace-Tang** [comment]
LGTM

**Comment by YaoZengzeng** [comment]
LGTM as well

**Pull request merged**